.github/workflows/test-flux.yaml

```diff
@@ -0,0 +1,43 @@
+name: Test Flux Scheduler
+on:
+  pull_request: []
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    permissions:
+      packages: read
+    strategy:
+      fail-fast: false
+      matrix:
+        container: ['fluxrm/flux-sched:focal']
+
+    container:
+      image: ${{ matrix.container }}
+      options: "--platform=linux/amd64 --user root -it"
+
+    name: ${{ matrix.container }}
+    steps:
+      - name: Make Space
+        run: |        
+          rm -rf /usr/share/dotnet
+          rm -rf /opt/ghc
+
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Install Reframe
+        run: |
+          /bin/bash ./bootstrap.sh
+          python3 -m pip install pytest
+          export PATH=$PWD/bin:$PATH
+          which reframe
+
+      # Any additional examples added here will be tested
+      - name: Start Flux and Run Test
+        run: |
+          export PATH=$PWD/bin:$PATH
+          which reframe   
+          flux start reframe -c tutorials/flux -C tutorials/flux/settings.py -l
+          flux start reframe -c tutorials/flux -C tutorials/flux/settings.py --run
+          flux start python3 ./test_reframe.py --rfm-user-config=tutorials/flux/settings.py -vvvv unittests/test_schedulers.py
```

Comment:
Hmm, I see you're running only the scheduler unit tests here. I'm trying to run the full suite and I'm getting another error earlier on:

```
____________________________________________________________________________________________ test_remote_autodetect ____________________________________________________________________________________________

remote_exec_ctx = <unittests.conftest._ExecutionContext object at 0x7fc809c96070>

    def test_remote_autodetect(remote_exec_ctx):
        # All we can do with this test is to trigger the remote auto-detection
        # path; since we don't know what the remote user system is, we cannot test
        # if the topology is right.
        partition = test_util.partition_by_scheduler()
        if not partition:
            pytest.skip('job submission not supported')

>       autodetect.detect_topology()

unittests/test_autodetect.py:116:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
reframe/frontend/autodetect.py:246: in detect_topology
    part._processor = ProcessorInfo(_remote_detect(part))
reframe/frontend/autodetect.py:154: in _remote_detect
    _log_contents(job.stdout)
reframe/frontend/autodetect.py:36: in _log_contents
    getlogger().debug(f'--- {filename} ---\n'
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

filename = '/code/rfm.6mxhbs0h/rfm-detect-job.out'

    def _contents(filename):
        '''Return the contents of a file.'''

>       with open(filename) as fp:
E       FileNotFoundError: [Errno 2] No such file or directory: '/code/rfm.6mxhbs0h/rfm-detect-job.out'

reframe/frontend/autodetect.py:30: FileNotFoundError
```

Have you tried the full suite? Did it work for you?

Reply:
I have not - let me give that a spin!

Reply:
Ok, I found this. The problem is that you haven't implemented the `wait()` in `FluxJobScheduler` and the fact that Python didn't complain, as it had to since the method is supposed to abstract, is because you inherit your implementation from the `PbsScheduler`, thus here reframe calls `PbsScheduler`'s `wait()`. Since Flux is a scheduler of its own, I would suggest to simply inherit from the abstract `JobScheduler` class. It could be that somehow this has messed up with your other unit test failure, perhaps not. But we should better fix that first.

Reply:
OK reproduced! And working on it! :gear:

Reply:
okay so I took a look at wait - and actually I had checked the version in the PbsScheduler and I don't see any issue. The reason is that the wait function is just sleeping for different intervals and then running another poll, and checking for finished:

```python
    def wait(self, job):
        intervals = itertools.cycle([1, 2, 3])
        while not self.finished(job):
            self.poll(job)
            time.sleep(next(intervals))
```
And then finished it just checking for an exception and returning if the job is completed (and actually we implement this in flux to check for finished based on the job state

```python
    def finished(self, job):
        if job.exception:
            raise job.exception

        return job.state in ['COMPLETED', 'CANCELLED', 'SUSPENDED']```
```
I'm trying to dig into the failure you showed above - it looks like this raises the error:

```python
autodetect.detect_topology()
```
Will report back in a few minutes (hopefully) with what is going on - I think I just found the part that is erroring out on the missing file! (I love debugging!) <3

Reply:
okay confirmed the erroneous "part" is indeed flux!
```
{
  "name": "default",
  "descr": "",
  "scheduler": "flux",
  "launcher": "local",
  "access": [],
  "container_platforms": [],
  "modules": [],
  "variables": [],
  "environs": [
    "builtin"
  ],
  "max_jobs": 8,
  "resources": []
}
```
We found device information (but not processor information):

```python
part.devices
Out[8]: [<reframe.core.systems.DeviceInfo at 0x7f7bf56b03d0>]
```
And then most of the if logic is skipped, and we wind up entering the `elif detect_remote_systems` and this is the command that fails:

```
                with runtime.temp_environment(modules=modules, variables=vars):
                    part._processor = ProcessorInfo(_remote_detect(part))
```
So I think what's happening here is that I haven't providing enough info about processor information, so it erroneously is looking for remote information, and I'm guessing those files (that aren't found) are what might hold it? So I need to figure out how to get this populated I think?

```
part.processor.info
Out[26]: {}
```

Reply:
I'm also seeing these directories scattered in the root:
```
# ls rfm.9xoaersf/
bin  bootstrap.sh  reframe  requirements.txt  rfm-detect-job.sh
```
and it could be flux doesn't know how to make one? Or run a requirement to work with one?

Reply:
@vkarak can you give me a sense of what this check is doing, and what might be missing for flux?

Reply:
Ok, I think that I am getting closer to understanding what is happening. Let me explain first how jobs are submitted and managed in ReFrame. The `compile` and `run` pipeline stages (see `reframe.core.pipeline`) of a `reframe.core.pipeline.RegressionTest` create `reframe.core.scheduler.Job` instances for the corresponding test jobs. These contain all the information related to the submitted job (e.g. the script filename, submission parameters, the output files, the status, the jobid etc.). The jobs are managed (submitted, collected, cancelled etc.) by the `reframe.core.schedulers.JobScheduler` backends, which implement concretely a set of abstract methods for those operations. All these method take the job as an argument and they update the job status (e.g., the job's `state` or any other information that needs to be updated). The framework uses the various `Job` instances to get *any* information related to the job, whereas the `JobScheduler` is merely used for managing the jobs (submitting, cancelling etc.).

Now this is what happens in your case. ReFrame has a mechanism for auto-detecting the processor topology not only for the local partitions but also for remote partitions, so that your tests can use that information to configure themselves (e.g., set the thread numbers or the CPU affinities, set compilation options etc.). To do so for remote partitions, reframe submits a job that executes itself with the `--detect-host-topology` option from a temporary directory and uses that output (after caching it for future) to configure the processor information for the target partition. That unit test executes only for remote partitions (i.e., any partition with a non-local scheduler) and does exactly that. Now what I see in your code is that you never set `Job.stdout` and `Job.stderr` (they have their [default](https://github.com/reframe-hpc/reframe/blob/9f0d16742166af0482c52ca53a01e4e5cf86ce29/reframe/core/schedulers/__init__.py#L295-L296) values), but instead you set them in your local `fluxjob` variable during submit. Although the values match the default ones and are in the correct directory (see later on my suggestion on how to better implement this), it seems that the stdout or stderr files are not created after your job finishes (supposedly, after the `wait()` is called). So I don't know if you need to do something specifically for Flux to make those appear or if Flux deletes them, or if there is delay before appearing? But then I'm wondering how the example works for your tutorial example? (Note: to debug this, I suggest removing the `os.rmtree()` call in the `_copy_reframe.__exit__()` method). This is where I reached so far.

Back to the design of `Job` vs. `JobScheduler`, I would suggest the following:

- Do not inherit from `PbsScheduler` but just `JobScheduler` (even conceptually is wrong; Flux has a completely different mentality and besides you would be getting errors mentioning PBS 😄 ).
- Define a `_FluxJob` extending `Job` (as the other schedulers do) and keep there *all* the information related to the Flux, see it as a wrapper to the result `flux.job.JobSpecV1.from_command`. Also, all this machinery can move to `_FluxJob.__init__()` and leave in the `JobScheduler.submit()` the actual job submission through the Flux executor.

That would make the design more compatible with the concepts of the framework and hopefully will avoid us surprises in the future.

Reply:
This was really helpful! So I didn't look closely at the wrappers to sched.Job (and this is how self.completed/self.cancelled make sense). As I was rewriting that part, I realized that the stdout and stderr that I set aren't from the job - they are custom paths that I wrote based on the script because I didn't realize the base job class had already specified one. So the bug was just that flux didn't know about the correct files to write to.

And then for the next bug for the autodetect, it looks like the files for output/error are relative, which works given the context when it's defined, but flux seemed to be writing to the original PWD. So the fix there is to take the abspath of the self.workdir and provide Flux with the full, absolute paths!

Just ran all the tests locally and green? Will push soon and fingers crossed it reproduces!

Reply:
Ok pushed! I restored the sleep I had extended to the original value, and tried to remove the empty newlines I had added in autodetect.py. Fingers crossed! :crossed_fingers: